Forbid cleaning of tmp directories that can be used by an active mutation/merge.

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Forbid cleaning of tmp directories that can be used by an active mutation/merge.

Detailed description / Documentation draft:
Mutator cannot be run in parallel with cleaner,
since cleaner may remove partial result of a merge,
and chances of this will be increase if:
- temporary_directories_lifetime is pretty low,
- someone adjust system-wide clock.

And these will lead to missing columns in these part,
and the server will figure out that something is wrong only on DETACH/ATTACH.

*NOTE: tested manually*

alexey-milovidov commented:
> cleaner may remove partial result of a merge

Only happens if no files were written during `temporary_directories_lifetime` (24 hours by default).

alexey-milovidov commented on src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp:
```diff
@@ -1187,6 +1187,16 @@ MergeTreeData::MutableDataPartPtr MergeTreeDataMergerMutator::mutatePartToTempor
+    ///
+    /// And these will lead to missing columns in these part,
+    /// and the server will figure out that something is wrong only on DETACH/ATTACH.
+    std::unique_lock lock(data.clear_old_temporary_directories_mutex);
```
But it may prevent cleanup thread to run for unbounded time. And also it will prevent concurrent mutations.

azat replied:
Ok, done w/o lock (by introducing `MergeTreeDataMergerMutator::hasTemporaryPart()`)

alexey-milovidov reviewed (changes requested):
.

azat commented:
>Only happens if no files were written during temporary_directories_lifetime (24 hours by default).

Yes, but:
- someone may set it to something low
- someone may adjust system time (i.e. after hardware replacement at start time will be far away, but after a short period of time NTP will adjust it)

azat committed:
Rewrite MergeTreeData::clearOldTemporaryDirectories() to use early continue

azat committed:
Do not try to remove temporary paths that is currently in written by merge/mutation

v2: rebase against MergeTask
v3: rebase due to conflicts in src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp
v4:
- rebase due to conflicts in src/Storages/MergeTree/MergeTask.cpp
- drop common/scope_guard_safe.h (not used)

azat commented:
>Functional stateless tests (release, DatabaseReplicated) — fail: 1, passed: 3238, skipped: 117

```
2021-09-11 03:55:27 [e010d80598a5] 2021.09.11 03:55:26.451731 [ 30617 ] {42848bab-ffaf-43b9-aacf-d71f8dfe972e} <Error> executeQuery: Code: 27. DB::ParsingException: Cannot parse input: expected 'version: ' at end of stream. (CANNOT_PARSE_INPUT_ASSERTION_FAILED) (version 21.11.1.8053) (from [::1]:46266) (comment: '/usr/share/clickhouse-test/queries/0_stateless/01161_all_system_tables.sh') (in query: SELECT * FROM system.distributed_ddl_queue LIMIT 10000 FORMAT Null), Stack trace (when copying this message, always include the lines below):
```

Unrelated, but interesting, sure right now there are no such nodes (`for p in $(zk-shell 127.1:2182 --run-once 'ls /clickhouse/task_queue/ddl'); do zk-shell 127.1:2182 --run-once "json_cat /clickhouse/task_queue/ddl/$p" | grep -q version || echo $p; done`), but what it was?
Create node and put the node's content should be atomic, right?
@alesapin maybe it is some issue in keeper?

azat commented:
>Functional stateless tests (release, DatabaseReplicated) — fail: 1, passed: 3265, skipped: 117

`01161_all_system_tables`

```
2021-09-17 13:35:11 [87453c98cfd3] 2021.09.17 13:35:10.931973 [ 25922 ] {af4987c4-a280-4211-b228-464f626afe98} <Error> executeQuery: Code: 27. DB::ParsingException: Cannot parse input: expected 'version: ' at end of stream. (CANNOT_PARSE_INPUT_ASSERTION_FAILED) (version 21.11.1.8109) (from [::1]:58870) (comment: '/usr/share/clickhouse-test/queries/0_stateless/01161_all_system_tables.sh') (in query: SELECT * FROM system.distributed_ddl_queue LIMIT 10000 FORMAT Null), Stack trace (when copying this message, always include the lines below):
```

Looks like should be fixed by #29061, in particular https://github.com/ClickHouse/ClickHouse/pull/29061/files#diff-27b2fdac05cd0b4e165bebc7509be60eaeb37a260ffb35cc69f5614e3f81c3d8R223-R228 (since `asyncTryGet` does not throw on ZNONODE``), @tavplubix please correct me if I'm wrong.

azat commented:
@alexey-milovidov can you do a final look please?

azat commented:
>Functional stateless tests (release, DatabaseReplicated) — fail: 1, passed: 3265, skipped: 117

Fixed in #29061 (and it is already merged, anyway this PR should not be related to the failure of `01161_all_system_tables`)

azat commented:
>Conflicting files
>src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp

Rebased to fix conflicts.

azat commented:
>[Functional stateless tests (release, DatabaseReplicated) — Some tests restarted, fail: 8, passed: 3286, skipped: 117](https://clickhouse-test-reports.s3.yandex.net/28760/20bdda60084621a2589cc45303d0fb2de0f33f5a/functional_stateless_tests_(release,_databasereplicated).html#fail1)

#29462

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=28760) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=28760) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=28760) it.</sub>

azat commented:
>Functional stateless tests (release, wide parts enabled) — fail: 1, passed: 3463, skipped: 2
>Functional stateless tests (thread) — fail: 1, passed: 3438, skipped: 27

- #30134 

>Stress test (thread) — Test script failed

```
Code: 241. DB::Exception: Received from localhost:9000. DB::Exception: Memory limit (total) exceeded: would use 120.69 GiB (attempt to allocate chunk of 2105374 bytes), maximum: 94.44 GiB. (MEMORY_LIMIT_EXCEEDED)
```

- #30218 is required to print queries

alexey-milovidov reviewed (approved):
LGTM
